File: cli/internal/agentauthz/auth_blob_test.go
package agentauthz

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"harness-data/cli/internal/harness"
)

const testBlob = "qdm1enc.testblob"

func TestResolveAuthBlobLocalFallbackDisabled(t *testing.T) {
	allow := false
	_, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: t.TempDir(),
		Config: harness.AuthzConfig{
			AllowLocalBlob: &allow,
		},
		Env: map[string]string{
			EnvAuthBlob:   testBlob,
			EnvAuthUserID: "local-user",
		},
	})
	if err == nil {
		t.Fatal("expected local fallback disabled error")
	}
}

func TestResolveAuthBlobReadsEnvBlob(t *testing.T) {
	resolved, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: t.TempDir(),
		Config:      harness.AuthzConfig{},
		Env: map[string]string{
			EnvAuthBlob:   testBlob,
			EnvAuthUserID: "env-user",
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if resolved.Source != BlobSourceEnv || resolved.Blob != testBlob || resolved.UserID != "env-user" {
		t.Fatalf("unexpected resolved env blob: %+v", resolved)
	}
}

func TestResolveAuthBlobReadsEnvBlobFile(t *testing.T) {
	root := t.TempDir()
	blobPath := filepath.Join(root, "admin-auth.blob")
	if err := os.WriteFile(blobPath, []byte(testBlob+"\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	resolved, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: root,
		Config:      harness.AuthzConfig{},
		Env: map[string]string{
			EnvAuthBlobFile: blobPath,
			EnvAuthUserID:   "env-user",
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if resolved.Source != BlobSourceEnvFile || resolved.Blob != testBlob || resolved.UserID != "env-user" {
		t.Fatalf("unexpected resolved env-file blob: %+v", resolved)
	}
}

func TestResolveAuthBlobReadsConfiguredFile(t *testing.T) {
	root := t.TempDir()
	if err := os.MkdirAll(filepath.Join(root, "config"), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(root, "config", "dev-auth.blob"), []byte(testBlob+"\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	resolved, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: root,
		Config: harness.AuthzConfig{
			BlobFile:  "config/dev-auth.blob",
			DevUserID: "local-user",
		},
		Env: map[string]string{},
	})
	if err != nil {
		t.Fatal(err)
	}
	if resolved.Source != BlobSourceFile || resolved.Blob != testBlob || resolved.UserID != "local-user" {
		t.Fatalf("unexpected configured blob: %+v", resolved)
	}
}

func TestResolveAuthBlobEnvBlobPrioritizedOverConfigFile(t *testing.T) {
	root := t.TempDir()
	if err := os.MkdirAll(filepath.Join(root, "config"), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(root, "config", "dev-auth.blob"), []byte("qdm1enc.fileblob\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	resolved, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: root,
		Config: harness.AuthzConfig{
			BlobFile:  "config/dev-auth.blob",
			DevUserID: "file-user",
		},
		Env: map[string]string{
			EnvAuthBlob:   testBlob,
			EnvAuthUserID: "env-user",
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if resolved.Source != BlobSourceEnv || resolved.Blob != testBlob || resolved.UserID != "env-user" {
		t.Fatalf("expected env blob to take priority: %+v", resolved)
	}
}

func TestResolveAuthBlobNoBlobAvailable(t *testing.T) {
	_, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: t.TempDir(),
		Config:      harness.AuthzConfig{},
		Env:         map[string]string{},
	})
	if err == nil {
		t.Fatal("expected no blob available error")
	}
}

func TestResolveAuthBlobRejectsGroupReadableFile(t *testing.T) {
	root := t.TempDir()
	blobPath := filepath.Join(root, "admin-auth.blob")
	if err := os.WriteFile(blobPath, []byte(testBlob+"\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	_, err := ResolveAuthBlob(ResolveOptions{
		ProjectRoot: root,
		Config:      harness.AuthzConfig{},
		Env: map[string]string{
			EnvAuthBlobFile: blobPath,
			EnvAuthUserID:   "env-user",
		},
	})
	if err == nil || !strings.Contains(err.Error(), "permissions must be 0600") {
		t.Fatalf("expected insecure permission error, got %v", err)
	}
}

File: packages/data-harness-cli/test/auth-blob.test.js
import assert from "node:assert/strict";
import { chmodSync, mkdirSync, mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import test from "node:test";

import {
  BLOB_SOURCE_ENV,
  BLOB_SOURCE_ENV_FILE,
  BLOB_SOURCE_FILE,
  resolveAuthBlob,
} from "../src/lib/authz/auth-blob.js";
import { ENV_AUTH_BLOB, ENV_AUTH_BLOB_FILE, ENV_AUTH_USER_ID } from "../src/lib/authz/constants.js";

const testBlob = "qdm1enc.testblob";

function tempRoot() {
  return mkdtempSync(path.join(tmpdir(), "auth-blob-"));
}

test("resolveAuthBlob local fallback disabled", () => {
  assert.throws(
    () =>
      resolveAuthBlob({
        projectRoot: tempRoot(),
        config: { allowLocalBlob: false },
        env: { [ENV_AUTH_BLOB]: testBlob, [ENV_AUTH_USER_ID]: "local-user" },
      }),
    /local blob fallback is disabled/,
  );
});

test("resolveAuthBlob reads env blob", () => {
  const resolved = resolveAuthBlob({
    projectRoot: tempRoot(),
    config: {},
    env: { [ENV_AUTH_BLOB]: testBlob, [ENV_AUTH_USER_ID]: "env-user" },
  });
  assert.equal(resolved.source, BLOB_SOURCE_ENV);
  assert.equal(resolved.blob, testBlob);
  assert.equal(resolved.userId, "env-user");
});

test("resolveAuthBlob reads env blob file", () => {
  const root = tempRoot();
  const blobPath = path.join(root, "admin-auth.blob");
  writeFileSync(blobPath, `${testBlob}\n`, { mode: 0o600 });
  chmodSync(blobPath, 0o600);
  const resolved = resolveAuthBlob({
    projectRoot: root,
    config: {},
    env: { [ENV_AUTH_BLOB_FILE]: blobPath, [ENV_AUTH_USER_ID]: "env-user" },
  });
  assert.equal(resolved.source, BLOB_SOURCE_ENV_FILE);
  assert.equal(resolved.blob, testBlob);
  assert.equal(resolved.userId, "env-user");
});

test("resolveAuthBlob reads configured file", () => {
  const root = tempRoot();
  mkdirSync(path.join(root, "config"), { recursive: true });
  const blobPath = path.join(root, "config", "dev-auth.blob");
  writeFileSync(blobPath, `${testBlob}\n`, { mode: 0o600 });
  chmodSync(blobPath, 0o600);
  const resolved = resolveAuthBlob({
    projectRoot: root,
    config: { blobFile: "config/dev-auth.blob", devUserId: "local-user" },
    env: {},
  });
  assert.equal(resolved.source, BLOB_SOURCE_FILE);
  assert.equal(resolved.blob, testBlob);
  assert.equal(resolved.userId, "local-user");
});

test("resolveAuthBlob env blob prioritized over config file", () => {
  const root = tempRoot();
  mkdirSync(path.join(root, "config"), { recursive: true });
  const blobPath = path.join(root, "config", "dev-auth.blob");
  writeFileSync(blobPath, "qdm1enc.fileblob\n", { mode: 0o600 });
  chmodSync(blobPath, 0o600);
  const resolved = resolveAuthBlob({
    projectRoot: root,
    config: { blobFile: "config/dev-auth.blob", devUserId: "file-user" },
    env: { [ENV_AUTH_BLOB]: testBlob, [ENV_AUTH_USER_ID]: "env-user" },
  });
  assert.equal(resolved.source, BLOB_SOURCE_ENV);
  assert.equal(resolved.blob, testBlob);
  assert.equal(resolved.userId, "env-user");
});

test("resolveAuthBlob no blob available", () => {
  assert.throws(
    () => resolveAuthBlob({ projectRoot: tempRoot(), config: {}, env: {} }),
    /no encrypted blob is available/,
  );
});

test("resolveAuthBlob rejects group readable file", { skip: process.platform === "win32" }, () => {
  const root = tempRoot();
  const blobPath = path.join(root, "admin-auth.blob");
  writeFileSync(blobPath, `${testBlob}\n`, { mode: 0o644 });
  chmodSync(blobPath, 0o644);
  assert.throws(
    () =>
      resolveAuthBlob({
        projectRoot: root,
        config: {},
        env: { [ENV_AUTH_BLOB_FILE]: blobPath, [ENV_AUTH_USER_ID]: "env-user" },
      }),
    /permissions must be 0600/,
  );
});
